src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.PropertyNameSet.cs

```diff
@@ -0,0 +1,145 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    public sealed partial class JsonDocument
+    {
+        // Ensure this stays on the stack by making it a ref struct.
+        private ref struct PropertyNameSet : IDisposable
+        {
+            // Data structure to track property names in an object while deserializing
+            // into a JsonDocument and validate that there are no duplicates. A small inline
+            // array is used for the first few property names and then a pooled array is used
+            // for the rest. Escaped property names are always stored in the pooled array.
+
+            private ReadOnlyMemory<byte>[]? _heapCache;
+            private int _heapCacheCount = 0;
+
+#if NET
+            private const int StackCacheThreshold = 16;
+
+            [InlineArray(StackCacheThreshold)]
+            private struct InlineRangeArray16
+            {
+                private Range _element0;
+            }
+
+            private InlineRangeArray16 _stackCache;
+            private int _stackCacheCount = 0;
+#endif
+
+            public PropertyNameSet()
+            {
+            }
+
+            internal void AddPropertyName(JsonProperty property, JsonDocument document)
+            {
+                DbRow dbRow = document._parsedData.Get(property.Value.MetadataDbIndex - DbRow.Size);
+                Debug.Assert(dbRow.TokenType is JsonTokenType.PropertyName);
+
+                // Name without quotes
+                ReadOnlyMemory<byte> utf8Json = document._utf8Json;
+                ReadOnlyMemory<byte> propertyName = utf8Json.Slice(dbRow.Location, dbRow.SizeOrLength);
+
+                if (dbRow.HasComplexChildren)
+                {
+                    propertyName = JsonReaderHelper.GetUnescaped(propertyName.Span);
+                }
+
+#if NET
+                for (int i = 0; i < _stackCacheCount; i++)
+                {
+                    ReadOnlySpan<byte> previousPropertyName = utf8Json[_stackCache[i]].Span;
+
+                    if (previousPropertyName.SequenceEqual(propertyName.Span))
+                    {
+                        ThrowHelper.ThrowJsonException_DuplicatePropertyNotAllowed(propertyName.Span);
+                    }
+                }
+#endif
+
+                // Positive heap cache count implies it is not null.
+                Debug.Assert(_heapCacheCount is 0 || _heapCache is not null);
+
+                for (int i = 0; i < _heapCacheCount; i++)
```

Comment:
Am I incorrect in thinking that validation is $\mathcal O(n^2)$ for objects of $n$ properties? That would make this code susceptible to potential DOS attacks. For properties exceeding the fixed threshold we should use a regular hashtable so that complexity remains at $\mathcal O(n)$.